# **Kwanza Tukule Data Analysis**

This report showcases the completion of the Kwanza Tukule Data Analyst Assessment, designed to evaluate my technical, analytical, and problem-solving skills. Using the provided anonymized sales dataset, I applied data cleaning, preparation, and exploratory analysis techniques to uncover actionable insights. The analysis includes identifying trends, customer segmentation, forecasting, and anomaly detection to address key business challenges. Strategic recommendations are derived from the findings, focusing on product performance, customer retention, and operational efficiency. Additionally, I developed a dashboard to visualize key metrics, enabling better decision-making. This report demonstrates my ability to analyze data effectively and present meaningful insights in a structured, professional manner.

### **Importing Libraries**

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import plotly.express as px
from matplotlib import pyplot as plt
%matplotlib inline


#html export
import plotly.io as pio
pio.renderers.default = 'notebook'

### **Loading Dataset**

In [ ]:
df = pd.read_excel(r'Case Study Data - Read Only.xlsx')
df

DATE ANONYMIZED CATEGORY ANONYMIZED PRODUCT  \
0      2024-08-18 21:32:00        Category-106       Product-21f4   
1      2024-08-18 21:32:00        Category-120       Product-4156   
2      2024-08-18 21:32:00        Category-121       Product-49bd   
3      2024-08-18 21:32:00         Category-76       Product-61dd   
4      2024-08-18 21:32:00        Category-119       Product-66e0   
...                    ...                 ...                ...   
333400 2024-01-09 20:49:00        Category-119       Product-e98d   
333401 2024-08-20 20:24:00        Category-106       Product-21f4   
333402 2024-08-20 20:24:00         Category-76       Product-6e9c   
333403 2024-10-10 18:19:00         Category-91       Product-523e   
333404 2024-10-10 18:19:00         Category-75       Product-b31e   

       ANONYMIZED BUSINESS ANONYMIZED LOCATION  QUANTITY  UNIT PRICE  
0            Business-de42       Location-1ba8         1       850.0  
1            Business-de42       Location-1ba8         2      1910.0  
2            Business-de42       Location-1ba8         1      3670.0  
3            Business-de42       Location-1ba8         1      2605.0  
4            Business-de42       Location-1ba8         5      1480.0  
...                    ...                 ...       ...         ...  
333400       Business-f9ff       Location-1979         1      1770.0  
333401       Business-72bc       Location-689f         1       850.0  
333402       Business-72bc       Location-689f         1      2575.0  
333403       Business-2693       Location-b27b         1      1520.0  
333404       Business-2693       Location-b27b         1      4420.0  

[333405 rows x 7 columns]

#### Attributes

1. **DATE:** Represents the date on which the transaction or activity occurred.  

2. **ANONYMIZED CATEGORY:** Categorization of the product or service involved in the transaction, with the actual category names replaced by anonymized labels. 

3. **ANONYMIZED PRODUCT:** Refers to the specific product associated with the transaction, with product names anonymized for confidentiality. 

4. **ANONYMIZED BUSINESS:** Represents the business entity involved in the transaction. 

5. **ANONYMIZED LOCATION:** The location where the transaction or activity occurred. 

6. **QUANTITY:** The quantity of the product involved in the transaction, recorded as an integer. 

7. **UNIT PRICE:** The price per unit of the product. 

### **Data Cleaning and Preparation**

In [8]:
# make column names and values uniform
df.columns = df.columns.str.lower().str.replace(' ', '_')

categorical_columns = df.dtypes[df.dtypes == 'object'].index
for c in categorical_columns:
    df[c] = df[c].str.lower().str.replace(' ', '_')

In [10]:
df

date anonymized_category anonymized_product  \
0      2024-08-18 21:32:00        category-106       product-21f4   
1      2024-08-18 21:32:00        category-120       product-4156   
2      2024-08-18 21:32:00        category-121       product-49bd   
3      2024-08-18 21:32:00         category-76       product-61dd   
4      2024-08-18 21:32:00        category-119       product-66e0   
...                    ...                 ...                ...   
333400 2024-01-09 20:49:00        category-119       product-e98d   
333401 2024-08-20 20:24:00        category-106       product-21f4   
333402 2024-08-20 20:24:00         category-76       product-6e9c   
333403 2024-10-10 18:19:00         category-91       product-523e   
333404 2024-10-10 18:19:00         category-75       product-b31e   

       anonymized_business anonymized_location  quantity  unit_price  
0            business-de42       location-1ba8         1       850.0  
1            business-de42       location-1ba8         2      1910.0  
2            business-de42       location-1ba8         1      3670.0  
3            business-de42       location-1ba8         1      2605.0  
4            business-de42       location-1ba8         5      1480.0  
...                    ...                 ...       ...         ...  
333400       business-f9ff       location-1979         1      1770.0  
333401       business-72bc       location-689f         1       850.0  
333402       business-72bc       location-689f         1      2575.0  
333403       business-2693       location-b27b         1      1520.0  
333404       business-2693       location-b27b         1      4420.0  

[333405 rows x 7 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 333405 entries, 0 to 333404
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   date                 333405 non-null  datetime64[ns]
 1   anonymized_category  333405 non-null  object        
 2   anonymized_product   333405 non-null  object        
 3   anonymized_business  333405 non-null  object        
 4   anonymized_location  333405 non-null  object        
 5   quantity             333405 non-null  int64         
 6   unit_price           333397 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(4)
memory usage: 17.8+ MB


In [14]:
# duplicates
df.duplicated().sum()

3524

There are 3524 duplicates. We need to drop them

In [16]:
#dropping duplicates
df = df.drop_duplicates()

In [17]:
df.isna().sum()

date                   0
anonymized_category    0
anonymized_product     0
anonymized_business    0
anonymized_location    0
quantity               0
unit_price             8
dtype: int64

The column 'unit_price' has 8 null values

In [20]:
df[df.unit_price.isna()]

date anonymized_category anonymized_product  \
108112 2024-07-03 17:53:00         category-94       product-3d7f   
150961 2024-12-16 18:33:00         category-79       product-dfc8   
151142 2024-12-22 14:42:00        category-122       product-15e0   
272379 2024-06-27 12:15:00         category-92       product-ccbc   
278284 2024-08-14 21:09:00        category-101       product-84a5   
278384 2024-12-30 14:17:00         category-95       product-15f3   
310385 2024-03-31 14:03:00        category-114       product-9204   
327152 2024-08-13 16:20:00        category-107       product-7eed   

       anonymized_business anonymized_location  quantity  unit_price  
108112       business-4fce       location-f37d         2         NaN  
150961       business-8bbf       location-3fc0         1         NaN  
151142       business-c575       location-1979         3         NaN  
272379       business-14b6       location-1979         1         NaN  
278284       business-4be1       location-bb69        21         NaN  
278384       business-1a74       location-f37d         1         NaN  
310385       business-c9dc       location-689f         1         NaN  
327152       business-0d61       location-1ba8         1         NaN

Lets try to see whether its the anonymized_product that determines the unit price

In [26]:
df_product_3d7fdf = df[df['anonymized_product'] == 'product-3d7f']
df_product_3d7fdf

date anonymized_category anonymized_product  \
196    2024-09-25 13:05:00         category-94       product-3d7f   
287    2024-11-13 19:36:00         category-94       product-3d7f   
319    2024-11-01 16:54:00         category-94       product-3d7f   
372    2024-07-11 11:20:00         category-94       product-3d7f   
447    2024-12-09 19:35:00         category-94       product-3d7f   
...                    ...                 ...                ...   
332771 2024-10-15 20:01:00         category-94       product-3d7f   
332802 2024-10-21 19:04:00         category-94       product-3d7f   
332842 2024-09-20 13:43:00         category-94       product-3d7f   
333002 2024-10-08 17:15:00         category-94       product-3d7f   
333188 2024-11-08 21:20:00         category-94       product-3d7f   

       anonymized_business anonymized_location  quantity  unit_price  
196          business-cdc1       location-e2f8         1       860.0  
287          business-7f77       location-8959         1       860.0  
319          business-2b91       location-3fc0         1       860.0  
372          business-2b24       location-66f4         5       875.0  
447          business-1b52       location-7f37         1       885.0  
...                    ...                 ...       ...         ...  
332771       business-0ea3       location-4ea1         1       860.0  
332802       business-5415       location-128a         1       860.0  
332842       business-5760       location-689f         1       860.0  
333002       business-8603       location-b27b         1       860.0  
333188       business-8bbf       location-3fc0         1       860.0  

[3164 rows x 7 columns]

In [30]:
df_product_3d7fdf['unit_price'].unique()

array([860., 875., 885., 870.,  nan])

For product 'product-3d7f' we can see there's 3 different unit prices: 860.0, 875.0, 885.0, 870.0

In [32]:
df_product_3d7fdf[df_product_3d7fdf['unit_price'] == 860.0]

date anonymized_category anonymized_product  \
196    2024-09-25 13:05:00         category-94       product-3d7f   
287    2024-11-13 19:36:00         category-94       product-3d7f   
319    2024-11-01 16:54:00         category-94       product-3d7f   
565    2024-09-15 15:52:00         category-94       product-3d7f   
780    2024-10-03 15:08:00         category-94       product-3d7f   
...                    ...                 ...                ...   
332771 2024-10-15 20:01:00         category-94       product-3d7f   
332802 2024-10-21 19:04:00         category-94       product-3d7f   
332842 2024-09-20 13:43:00         category-94       product-3d7f   
333002 2024-10-08 17:15:00         category-94       product-3d7f   
333188 2024-11-08 21:20:00         category-94       product-3d7f   

       anonymized_business anonymized_location  quantity  unit_price  
196          business-cdc1       location-e2f8         1       860.0  
287          business-7f77       location-8959         1       860.0  
319          business-2b91       location-3fc0         1       860.0  
565          business-f749       location-3fc0         2       860.0  
780          business-cdc1       location-e2f8         1       860.0  
...                    ...                 ...       ...         ...  
332771       business-0ea3       location-4ea1         1       860.0  
332802       business-5415       location-128a         1       860.0  
332842       business-5760       location-689f         1       860.0  
333002       business-8603       location-b27b         1       860.0  
333188       business-8bbf       location-3fc0         1       860.0  

[2344 rows x 7 columns]

Even for product 'product-3d7f' with a specific unit price of 860.0, we still have no supporting data to associate that price to. <br> <br>
Hence, its safe to drop the nulls

In [34]:
# dropping nulls
df = df.dropna()
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 329873 entries, 0 to 333404
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   date                 329873 non-null  datetime64[ns]
 1   anonymized_category  329873 non-null  object        
 2   anonymized_product   329873 non-null  object        
 3   anonymized_business  329873 non-null  object        
 4   anonymized_location  329873 non-null  object        
 5   quantity             329873 non-null  int64         
 6   unit_price           329873 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(4)
memory usage: 20.1+ MB


In [ ]:
# checking unique values for each column
for col in df.columns:
    print(col)
    print(df[col].unique()[:5]) # first 5 unique values
    print(df[col].nunique())  # distinct values in our dataset
    print()
    

date
<DatetimeArray>
['2024-08-18 21:32:00', '2024-08-06 19:36:00', '2024-06-23 19:37:00',
 '2024-11-25 16:42:00', '2024-09-06 08:42:00']
Length: 5, dtype: datetime64[ns]
96702

anonymized_category
['category-106' 'category-120' 'category-121' 'category-76' 'category-119']
46

anonymized_product
['product-21f4' 'product-4156' 'product-49bd' 'product-61dd'
 'product-66e0']
817

anonymized_business
['business-de42' 'business-7488' 'business-2460' 'business-aab8'
 'business-f13b']
4800

anonymized_location
['location-1ba8' 'location-4ea1' 'location-128a' 'location-bb69'
 'location-1979']
53

quantity
[1 2 5 4 3]
79

unit_price
[ 850. 1910. 3670. 2605. 1480.]
1050



### **Summary of Issues Identified and Steps Taken to Resolve Them:**
1. **Duplicate Records:**
- *Issue Identified:* The dataset initially contained 3524 duplicate records, which could skew the analysis and affect the accuracy of any insights derived from the data.
- *Steps Taken:* The duplicates were identified and dropped from the dataset using the .duplicated().sum() method and .drop_duplicates() to ensure each record is unique and accurately represents individual sales transactions.
2. **Null Values in 'unit_price' Column:**
- *Issue Identified:* There were 8 null values in the unit_price column, which is critical as the price of the product must be available for proper calculations of sales value and other analyses.
- *Steps Taken:* The null values in the unit_price column were identified using .isnull().sum() and subsequently dropped from the dataset using .dropna() to remove any incomplete records that might lead to errors in subsequent analysis.<br><br>

**Final Dataset:**
After addressing these issues, the dataset now contains 329,873 non-null records, with no duplicate entries and all null values removed from the unit_price column. The dataset's columns now have consistent data types and no missing or redundant data, making it ready for further analysis.
